### ISSUE TITLE
Proposed design for public AMQP queueing API

### ISSUE THREAD

**Issue by preeth1** [replicate/cog#404]
# Context
We currently have a private queue-based redis API that replicate.com uses to talk to Cog models, and we want to turn that into a well-defined public API that uses standard protocols (in this case AMQP). 

# Requirements
- Run predictions from JSON inputs in a request queue
- Put JSON output of prediction on a response queue specified in request
- Read input files from a URL (HTTP/S3/GCS) and write output files to particular location in an S3/GCS bucket
- Record log output of prediction, in some way that can be read as the prediction is running
- Put in-progress outputs from the model onto the response queue, as the model is running
- Validate inputs and outputs based on a JSON Schema
- Include metrics about run in response: model start up time, request processing time, and system metrics like CPU, RAM, and GPU memory used.
- Errors and logs in response if prediction fails.
- Cancel a running prediction.

# Strawman design

At a high-level, the AMQP API uses the same request and response objects as the HTTP API, defined with an OpenAPI schema. But, instead of HTTP requests and responses, these objects are sent over AMQP request and response queues.

## Requests

- The message body is the same format as the HTTP API request body being implemented in #378. It contains the input and any other metadata.
- Message body encoded as JSON (`content_type` property set to `application/json`).
- The response queue is specified with the `reply_to` property on the message.
- Files are represented as URIs, prefixed with either `file://`, `http://`, `https://`, `s3://`, or `gs://`.
- Example request body:

```json
{
  "input": {
    "image": "s3://hooli-input-bucket/hotdog.jpg",
    "disentanglement_threshold": 5
  },
  "output_file_prefix": "s3://hooli-output-bucket/5245/"
}
```

## Responses

- The message body is the same format as the HTTP API response body. It contains the status, output, logs, metrics, and other metadata.
- Files are represented as URIs, like in the response. The model will upload them to `output_file_prefix` defined in the request. Credentials are provided at runtime via environment variables.
- The response will be sent to the `reply_to` queue specified in the request.
- Message body encoded as JSON (`content_type` property set to `application/json`).
- Includes any message IDs present in the request (`message_id`, `correlation_id` properties).
- Includes timing information and other performance metrics (setup time, prediction time, CPU time, memory usage, etc).
- Includes logs. For each log message, or some log message updating frequency, this message will be sent multiple times, with `status: "processing"`.
- For models that return in-progress output as as they are running, this message will be sent multiple times, with `status: "processing"` and the "output" key set to the output.
- Example request body: 

```json
{
  "status": "success",
  "output": "s3://hooli-output-bucket/5245/output.jpg",
  "metrics": {
    "setup_time": <time>,
    "prediction_time": <time>
  },
  "logs": <array of strings, one per line>
}
```

## Schema

The AMQP API is specified with a subset of the OpenAPI/JSON Schema used for the HTTP API.

It is available both as a `org.cogmodel.openapi_schema` label on the image, and at `/openapi.json` in the HTTP API.

The relevant subset looks like this:

```json
{
  "components": {
    "schemas": {
      "Input": {
        "title": "Input",
        "required": ["image", "disentanglement_threshold"],
        "type": "object",
        "properties": {
          "image": {
            "description": "Image to manipulate",
            "type": "string",
            "format": "uri",
            "x-order": 0
          },
          "disentanglement_threshold": {
            "type": "number",
            "x-order": 1
          }
        }
      },
      "Output": {
        "title": "Output",
        "type": "string",
        "format": "uri"
      }
    }
  }
}

```

# Big differences between old and new API
- New system would use a well-defined, public AMQP API instead of the current system of using redis queues, to run predictions. Rationale: want to use standard tools, and make it and easy to run predictions in other workflows.
- Old system had separate queues for responses, logs, and timing info, new system puts all of that info in the response queue. Rationale: easier to keep track of 1 queue?

# Todo
- Cancel running predictions
- Error handling
- Pass credentials for file downloads/uploads with a request for scoping
- Backwards compatibility on Replicate

# Some references worth reading:
- [https://www.cloudamqp.com/blog/rpc-with-rabbitmq-direct-reply-to.html](https://www.cloudamqp.com/blog/rpc-with-rabbitmq-direct-reply-to.html)
- [https://www.rabbitmq.com/publishers.html#message-properties](https://www.rabbitmq.com/publishers.html#message-properties)
- Example implementation: [https://www.ribice.ba/golang-rabbitmq-client/](https://www.ribice.ba/golang-rabbitmq-client/)

Edited by @preeth1, @bfirsh, ... (maintainers -- add your name here! consider this is a wiki.)

**Comment by CtrlC-Root** [comment]
The requirements you have listed are mostly a superset of what we need at Rosebud.

> # Requirements
>     * Run predictions from JSON inputs in a request queue
>     * Put JSON output of prediction on a response queue specified in request
>     * Read input files from a URL (HTTP/S3/GCS) and write output files to particular location in an S3/GCS bucket
>     * Record log output of prediction, in some way that can be read as the prediction is running

I'm not sure that it's worth doing this in-band but maybe allowing some way to specify the location of the logs at least might be useful. Depending on how much you log this could be a lot of data and it would only be useful for a developer (who is probably looking at the docker compose console output already). It wouldn't be useful for a non-technical person using a demo or an API client.

>     * Put in-progress outputs from the model onto the response queue, as the model is running
>     * Validate inputs and outputs based on a JSON Schema
>     * Include metrics about run in response: model start up time, request processing time, and system metrics like CPU, RAM, and GPU memory used.
>     * Errors and logs in response if prediction fails.

Ditto as above.

>     * Cancel a running prediction.

We do this only in response to the API client going away but doing it explicitly in-band might be nicer.

One additional thing I'll add is that we wanted to document what kind of GPUs and how much video memory models needed ahead of time. This would be useful if we were deploying these to k8s or somewhere where we could use that information to inform deployment decisions. As it happens we're just running on top of compute instances for now and even if we weren't we'd probably use the same kind of GPU (i.e. T4s) everywhere so it's not something we've bothered to add yet and don't depend on it.

> ## Requests
>     * The message body is the same format as the HTTP API request body being implemented in [Use Pydantic for type annotations and FastAPI/OpenAPI for HTTP API #378](https://github.com/replicate/cog/pull/378). It contains the input and any other metadata.
>     * Message body encoded as JSON (`content_type` property set to `application/json`).
>     * The response queue is specified with the `reply_to` property on the message.
>     * Files are represented as URIs, prefixed with either `file://`, `http://`, `https://`, `s3://`, or `gs://`.

What would the `file://` prefix be used for? Local files?

> ## Responses
>     * The message body is the same format as the HTTP API response body. It contains the status, output, logs, metrics, and other metadata.
>     * Files are represented as URIs, like in the response. The model will upload them to `output_file_prefix` defined in the request. Credentials are provided at runtime via environment variables.

I saw somewhere else mentioned that credentials would be included in the request. Regardless I think leaving this as unspecified as possible is ideal. If you're running on AWS, GCP, or Azure you likely want to have the option to fall back on the IAM role you've assigned the service or compute instance and control what the model has access to through that. With boto3 and the google SDKs you usually accomplish this by not specifying any credentials at all. As a fallback for other use cases where you might be using an external S3 API it's useful to be able to specify these as settings when starting the model runner too.

>     * The response will be sent to the `reply_to` queue specified in the request.
>     * Message body encoded as JSON (`content_type` property set to `application/json`).
>     * Includes any message IDs present in the request (`message_id`, `correlation_id` properties).

We use the `correlation_id` to match up response messages to request messages.

>     * Includes timing information and other performance metrics (setup time, prediction time, CPU time, memory usage, etc).
>     * Includes logs. For each log message, or some log message updating frequency, this message will be sent multiple times, with `status: "processing"`.

Ditto as what I mentioned about this in the requirements section above.

**Comment by evilstreak** [comment]
I think this looks sensible! I've made a few notes about my thoughts:

- Do we want to be able to terminate model containers via the queue? I've previously used messages like that for autoscaling. If the logic working out the appropriate pool size concluded the pool should drop from 7 workers to 5 workers, it would enqueue a couple of "terminate" messages. Whichever workers finished their jobs first would ack the message and shutdown. It allowed me to do graceful shutdown of workers without having to work out which containers were idle from outside the system. Another option is shutting down when there's no work left (after a timeout?) but that could make it hard to intentionally hold workers open to mitigate long startup times.

- Should the request include the name of another queue used for cancellations? That seems more flexible than having it specified when the model container is started. It'd also allow a runtime choice of whether to have a single queue for cancellations that all containers listen to, or a queue per request. Not sure about the performance implications of either of those choices. The reason I'm thinking a different queue from requests is that reusing the same queue for cancellations as for requests would mean running models would have to continually check the request queue, and mostly ignore what's there. Specifying a separate queue per model avoids needing to specify an identifier. Presumably we could use correlation_id to make shared queues work.

- Potentially useful status values, beyond `processing` and `success`:
    - the model run failed internally (`failed`?)
    - the model run was cancelled via the queue (`cancelled`?)
    - the model run was interrupted (eg spot instance shutdown) (`interrupted`?)

- @CtrlC-Root's response about logs is interesting. Is it worth putting those on a separate queue, or at least having messages structured in a way that makes it easy to determine whether the only new information is logs, so they can be ignored?

**Comment by andreasjansson** [comment]
I think this generally looks good.

At the moment we define a **new response queue per request**. Does the proposal above suggest we use Direct Reply-to, a single response queue for each model, or a response queue per request? For polling purposes it would probably be easiest to use the existing pattern of a response queue per request, but that is inefficient. Is there a design that makes it easy to poll individual request status while being efficient and reliable?

**Termination** is interesting. How about we define a separate termination queue per model, with the worker listening on the termination queue in a separate thread? When a message comes in on the termination queue it signals the main worker process to cancel.

Additional (future?) questions:
*  We've had several requests for **streaming inputs** (e.g. send a stream of audio to a model and get a text transcription back in real time). This might be a question for a future version, but could we fit that into the current design? Or would that feature require a complete redesign?
* How do we handle **retries**? This is particularly interesting for spot instance-backed architectures, where jobs may fail mid-prediction. It's not a big deal for fast models, but for slow ones we might want to consider an API for posting continuous checkpoints to be retried from.

**Comment by CtrlC-Root** [comment]
> For polling purposes it would probably be easiest to use the existing pattern of a response queue per request, but that is inefficient. Is there a design that makes it easy to poll individual request status while being efficient and reliable?

Generally subscribing to queues to automatically receive new messages and doing so in batches. Polling is simpler and faster to implement correctly which is why we started with it in our implementation. In the future we want to look at invoking models to run in parallel and in that case it probably would be convenient to use the same client-bound queue as the reply to queue for each run request but there's nothing in AMQP or RabbitMQ to prevent you from using multiple queues.

>     * We've had several requests for **streaming inputs** (e.g. send a stream of audio to a model and get a text transcription back in real time). This might be a question for a future version, but could we fit that into the current design? Or would that feature require a complete redesign?

We're talking internally about this too. For some of our use cases where we process videos we could spread the work out through the entire pipeline if we switch to working on batches of frames instead of entire videos. I think this is probably a much more complicated problem to solve though. Depending on the performance and scale you're aiming for you probably don't want to stream the actual data through RabbitMQ.

>     * How do we handle **retries**? This is particularly interesting for spot instance-backed architectures, where jobs may fail mid-prediction. It's not a big deal for fast models, but for slow ones we might want to consider an API for posting continuous checkpoints to be retried from.

Interesting. This would be useful for development or staging environments where you can back the models with spot instances to decrease operational cost. In our production environment I would assume even if a retry is successful it's likely to complete too late from the initial request to be useful for the end user. So it should probably be configurable so you can disable it entirely if needed.

**Comment by andreasjansson** [comment]
# Full snapshots vs incremental updates

The original proposal in this ticket describes an API where Cog's AMQP worker is sending the entire state to the response queue on each update (e.g. a new log line is produced, or intermediate outputs are generated). @evilstreak and I spoke in person about this design and contrasted it with a design where only incremental updates are sent to the response queue.

## Full snapshots

In this design, the response queue is fed with full snapshots that represent the cumulative state of the prediction up to that point in time. Those snapshots are put on the queue every time there is a change to the model. For example, the following predict function
```python
def predict(self):
    print("first log line")
    print("second log line")
    time.sleep(2)
    print("third log line")
    return "output"
```

would send the following messages to the queue (disregarding metrics):

1.
```json
{
  "status": "processing",
  "output": null,
  "logs": ["first log line"]
}
```

2.
```json
{
  "status": "processing",
  "output": null,
  "logs": ["first log line", "second log line"]
}
```

3.
```json
{
  "status": "processing",
  "output": null,
  "logs": ["first log line", "second log line", "third log line"]
}
```

4.
```json
{
  "status": "success",
  "output": {"text": "output"},  # <- exact format tbd
  "logs": ["first log line", "second log line", "third log line"]
}
```

Benefits:
* You can always get a full snapshot of a prediction
* No need for the infrastructure that hosts the Cog models to "merge" states

Downsides:
* State cumulatively increases in size
* Lots of duplicated data kept in RabbitMQ
* You might want to forward incremental updates, e.g. to your own logging system. In that case you'd have to manually compute diffs between subsequent response messages
* There’s a lot of output that might never get used if you're not interested in the logs or intermediate outputs
* This design intuitively felt surprising to both of us

## Incremental updates

Instead of sending a full cumulative snapshot of each prediction state to the response queue, we considered a design where we only send incremental updates of what has changed recently.

For efficiency we could batch log lines and only send logs at a fixed interval. Progressive outputs and final results would still be sent as soon as they are produced by the model.

The predict function above would produce the following messages:

1.
```json
{
  "status": "processing",
  "output": null,
  "logs": ["first log line", "second log line"] # <- batched
}
```

2.
```json
{
  "status": "processing",
  "output": null,
  "logs": ["third log line"] # <- only one new log line
}
```

3.
```json
{
  "status": "success",
  "output": {"text": "output"},
  "logs": []
}
```

Benefits:
* No duplicated data between messages
* Message sizes don't indefinitely grow larger depending on the number of messages
* Easier to forward logs to log processors

Downsides:
* User of Cog models may need to manually "merge" incremental updates into actual prediction states. In the example above, at timestep 2 `["third log line"]` would have to be concatenated with the existing log line array, and at timestep 3 the status would have to be updated to "success" and output overwritten by `{"text": "output"}`.

## Discussion

My preference is for the incremental updates option, for the reasoning outlined above.

**Comment by bfirsh** [comment]
A few thoughts:

1. The only field which seems to be special here is log messages. Most users of Cog don't care about logs in the model output, like @CtrlC-Root mentioned. Logs would usually be shipped off directly to logging systems from the model via K8S, Docker log adapters, etc.

    If most people don't care about getting their output live, logs should probably not be sent by default. We could have some option for enabling it.
  
    This is making me wonder if logs would actually work better as a separate type of message, or a separate queue, like we have at the moment? Or maybe even just some completely separate system? (Just ship to syslog?!)

2. One neat property of "full snapshots" is the AMQP API messages map 1-1 to the HTTP API. It would be a shame to lose this symmetry -- for documentation, schema, simplicity of writing clients, elegance of internals, etc.

**Comment by andreasjansson** [comment]
This issue is superseded by https://github.com/replicate/cog/issues/443 and the issues tracked inside it. Hopefully I've captured the discussions and critiques we've had on this strawman design.
